src/js/angular/security/controllers.js

```diff
@@ -338,40 +353,90 @@ securityModule.controller('CommonUserCtrl', ['$rootScope', '$scope', '$http', 't
                 || $scope.grantedAuthorities.WRITE_REPO[uniqueKey];
         };
 
+        $scope.hasGraphqlPermission = function (repository) {
+            const uniqueKey = createUniqueKey(repository);
+            return repository.id !== SYSTEM_REPO && $scope.grantedAuthorities.GRAPHQL['*']
+                || $scope.grantedAuthorities.GRAPHQL[uniqueKey];
+        };
+
         $scope.readCheckDisabled = function (repoOrWildCard) {
             return $scope.hasWritePermission(repoOrWildCard)
                 || repoOrWildCard.id !== SYSTEM_REPO && repoOrWildCard !== '*' && $scope.grantedAuthorities.READ_REPO['*']
                 || $scope.hasEditRestrictions();
         };
 
         $scope.writeCheckDisabled = function (repoOrWildCard) {
-            return $scope.userType === UserType.ADMIN || $scope.userType === UserType.REPO_MANAGER
+            return $scope.userType === UserType.ADMIN
+                || $scope.userType === UserType.REPO_MANAGER
                 || repoOrWildCard.id !== SYSTEM_REPO && repoOrWildCard !== '*' && $scope.grantedAuthorities.WRITE_REPO['*']
                 || $scope.hasEditRestrictions();
         };
 
+        /**
+         * Determines whether the GraphQL checkbox for a given repository or wildcard
+         * should be disabled.
+         *
+         * @param {Object|string} repoOrWildCard - A repo object or the string wildcard '*'
+         * @returns {boolean} true if disabled, false if enabled
+         */
+        $scope.graphqlCheckDisabled = function (repoOrWildCard) {
+            if ($scope.userType === UserType.ADMIN || $scope.userType === UserType.REPO_MANAGER) {
+                return true;
+            }
+
+            if ($scope.hasEditRestrictions()) {
+                return true;
+            }
+
+            if (repoOrWildCard !== '*' && $scope.grantedAuthorities.GRAPHQL['*']) {
+                return true;
+            }
+
+            if (repoOrWildCard === '*' && $scope.grantedAuthorities.GRAPHQL['*']) {
+                return false;
+            }
+
+            let uniqueKey;
+            if (repoOrWildCard !== null && typeof repoOrWildCard === 'object') {
+                uniqueKey = $scope.createUniqueKey(repoOrWildCard);
+            } else {
+                uniqueKey = repoOrWildCard;
+            }
+
+            const hasReadWildcard = $scope.grantedAuthorities.READ_REPO['*'];
+            const hasWriteWildcard = $scope.grantedAuthorities.WRITE_REPO['*'];
+            const hasReadForRepo   = $scope.grantedAuthorities.READ_REPO[uniqueKey];
+            const hasWriteForRepo  = $scope.grantedAuthorities.WRITE_REPO[uniqueKey];
+
+            return !(hasReadWildcard || hasWriteWildcard || hasReadForRepo || hasWriteForRepo);
+        };
+
+
         $scope.createUniqueKey = function (repository) {
             return createUniqueKey(repository);
         };
 
         $scope.userType = UserType.USER;
         $scope.grantedAuthorities = {
```

Comment:
I just wonder why you didn't create some model for these granted authorities and encapsulate all the stuff related to it there. It's just mindbending as it is now.